File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();

        // Initialize the output array with 1s
        std::vector<int> output(n, 1);

        // Calculate left product and store it directly in the output array
        int left_product = 1;
        for (int i = 0; i < n; ++i) {
            output[i] = left_product;
            left_product *= nums[i];
        }

        // Calculate right product and multiply it with the corresponding left product in the output array
        int right_product = 1;
        for (int i = n - 1; i >= 0; --i) {
            output[i] *= right_product;
            right_product *= nums[i];
        }

        return output;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l <= r:
            mid = (l + r) // 2
            if target == nums[mid]:
                return mid

            #left sorted portion
            if nums[l] <= nums[mid]:
                #target in right sorted portion
                if target > nums[mid] or target < nums[l]:
                    l = mid + 1
                else:
                    #target in left sorted portion
                    r = mid - 1
            else:
                #target in left sorted portion
                if target < nums[mid] or target > nums[r]:
                    r = mid - 1
                else:
                    #target in right sorted portion
                    l = mid + 1
        return -1

     